src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -0,0 +1,368 @@
+                InitiateUnload();
+            }
+        }
+
+        private void InitiateUnload()
+        {
+            var unloading = Unloading;
+            Unloading = null;
+            unloading?.Invoke(this);
+
+            // When in Unloading state, we are not supposed to be called on the finalizer
+            // as the native side is holding a strong reference after calling Unload
+            lock (_unloadLock)
+            {
+                if (!s_isProcessExiting)
+                {
+                    Debug.Assert(_state == InternalState.Alive);
+
+                    var thisStrongHandle = GCHandle.Alloc(this, GCHandleType.Normal);
+                    var thisStrongHandlePtr = GCHandle.ToIntPtr(thisStrongHandle);
+                    // The underlying code will transform the original weak handle
+                    // created by InitializeLoadContext to a strong handle
+                    PrepareForAssemblyLoadContextRelease(_nativeAssemblyLoadContext, thisStrongHandlePtr);
+                }
+
+                _state = InternalState.Unloading;
+            }
+
+            if (!s_isProcessExiting)
+            {
+                lock (s_contextsToUnload)
+                {
+                    s_contextsToUnload.Remove(_id);
+                }
+            }
+        }
+
+        // Event handler for resolving native libraries.
+        // This event is raised if the native library could not be resolved via
+        // the default resolution logic [including AssemblyLoadContext.LoadUnmanagedDll()]
+        // 
+        // Inputs: Invoking assembly, and library name to resolve
+        // Returns: A handle to the loaded native library
+        public event Func<Assembly, string, IntPtr> ResolvingUnmanagedDll;
+
+        // Event handler for resolving managed assemblies.
+        // This event is raised if the managed assembly could not be resolved via
+        // the default resolution logic [including AssemblyLoadContext.Load()]
+        // 
+        // Inputs: The AssemblyLoadContext and AssemblyName to be loaded
+        // Returns: The Loaded assembly object.
+        public event Func<AssemblyLoadContext, AssemblyName, Assembly> Resolving;
+
+        public event Action<AssemblyLoadContext> Unloading;
+
+        // Occurs when an Assembly is loaded
+        public static event AssemblyLoadEventHandler AssemblyLoad;
+
+        // Occurs when resolution of type fails
+        public static event ResolveEventHandler TypeResolve;
+
+        // Occurs when resolution of resource fails
+        public static event ResolveEventHandler ResourceResolve;
+
+        // Occurs when resolution of assembly fails
+        // This event is fired after resolve events of AssemblyLoadContext fails
+        public static event ResolveEventHandler AssemblyResolve;
+
+        public static AssemblyLoadContext Default => LazyInitializer.EnsureInitialized(ref s_defaultAssemblyLoadContext, () => new AppPathAssemblyLoadContext());
```

Comment:
The `Default` will be initialized pretty much every time ALC is used for anything. I do not think it is necessary to use LazyInitializer here.

It may be best to make this `public static AssemblyLoadContext Default => AppPathAssemblyLoadContext.s_default;` and have readonly field field in AppPathAssemblyLoadContext. 

We use similar pattern for encodings: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Text/Encoding.cs#L1177
https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Text/UTF8Encoding.cs#L63

Reply:
good idea!